File: android/src/main/java/it/near/sdk/reactnative/rnnearitsdk/RNNearItUtils.java
/*
 * Copyright (c) 2017 Mattia Panzeri <mattia.panzeri93@gmail.com>
 * Last changes by Federico Boschini <federico@nearit.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package it.near.sdk.reactnative.rnnearitsdk;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.util.Base64;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.google.gson.Gson;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedOutputStream;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.UnsupportedEncodingException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import it.near.sdk.reactions.contentplugin.model.Content;
import it.near.sdk.reactions.contentplugin.model.ContentLink;
import it.near.sdk.reactions.contentplugin.model.ImageSet;
import it.near.sdk.reactions.couponplugin.model.Claim;
import it.near.sdk.reactions.couponplugin.model.Coupon;
import it.near.sdk.reactions.customjsonplugin.model.CustomJSON;
import it.near.sdk.reactions.feedbackplugin.model.Feedback;
import it.near.sdk.reactions.simplenotificationplugin.model.SimpleNotification;
import it.near.sdk.recipes.inbox.model.HistoryItem;
import it.near.sdk.trackings.TrackingInfo;

import static it.near.sdk.reactnative.rnnearitsdk.RNNearItConstants.*;

@SuppressWarnings({"CharsetObjectCanBeUsed", "Convert2Diamond"})
class RNNearItUtils {

    private static final String TAG = "RNNearItUtils";

    static String bundleTrackingInfo(final TrackingInfo trackingInfo) throws Exception {
        // JSONify trackingInfo
        final String trackingInfoJson = new Gson().toJson(trackingInfo);

        // Encode to base64
        return Base64.encodeToString(trackingInfoJson.getBytes("UTF-8"), Base64.DEFAULT);
    }

    @Nullable
    static TrackingInfo unbundleTrackingInfo(final String trackingInfoBase64) {
        // Decode from base64
        try {
            final String trackingInfoJsonString = new String(Base64.decode(trackingInfoBase64, Base64.DEFAULT), "UTF-8");
            return new Gson().fromJson(trackingInfoJsonString, TrackingInfo.class);
        } catch (UnsupportedEncodingException e) {
            Log.e(TAG, "error while decoding trackingInfo");
        }

        return null;
    }

    static WritableMap bundleCoupon(final Coupon coupon) {
        final WritableMap couponMap = new WritableNativeMap();

        couponMap.putString(EVENT_CONTENT_TITLE, coupon.getTitle());
        couponMap.putString(EVENT_COUPON_DESCRIPTION, coupon.description);
        couponMap.putString(EVENT_COUPON_VALUE, coupon.value);
        couponMap.putString(EVENT_COUPON_EXPIRES_AT, coupon.expires_at);
        couponMap.putString(EVENT_COUPON_REDEEMABLE_FROM, coupon.redeemable_from);
        couponMap.putString(EVENT_COUPON_SERIAL, coupon.getSerial());
        couponMap.putString(EVENT_COUPON_CLAIMED_AT, coupon.getClaimedAt());
        couponMap.putString(EVENT_COUPON_REDEEMED_AT, coupon.getRedeemedAt());

        // Coupon icon handling
        if (coupon.getIconSet() != null) {
            couponMap.putMap(EVENT_IMAGE, bundleImageSet(coupon.getIconSet()));
        }

        return couponMap;
    }

    static Coupon unbundleCoupon(final ReadableMap bundledCoupon) {
        Coupon coupon = new Coupon();
        coupon.name = getNullableField(bundledCoupon, EVENT_CONTENT_TITLE);
        coupon.description = getNullableField(bundledCoupon, EVENT_COUPON_DESCRIPTION);
        coupon.value = getNullableField(bundledCoupon, EVENT_COUPON_VALUE);
        coupon.expires_at = getNullableField(bundledCoupon, EVENT_COUPON_EXPIRES_AT);
        coupon.redeemable_from = getNullableField(bundledCoupon, EVENT_COUPON_REDEEMABLE_FROM);
        List<Claim> claims = new ArrayList<Claim>();
        Claim claim = new Claim();
        claim.serial_number = getNullableField(bundledCoupon, EVENT_COUPON_SERIAL);
        claim.claimed_at = getNullableField(bundledCoupon, EVENT_COUPON_CLAIMED_AT);
        claim.redeemed_at = getNullableField(bundledCoupon, EVENT_COUPON_REDEEMED_AT);
        claims.add(claim);
        coupon.claims = claims;
        if (bundledCoupon.hasKey(EVENT_IMAGE)) {
            ReadableMap imageSet = bundledCoupon.getMap(EVENT_IMAGE);
            coupon.setIconSet(unbundleImageSet(imageSet));
        }
        return coupon;
    }

    static WritableMap bundleContent(final Content content) {
        final WritableMap bundledContent = new WritableNativeMap();

        bundledContent.putString(EVENT_CONTENT_TITLE, content.title);
        bundledContent.putString(EVENT_CONTENT_TEXT, content.contentString);

        if (content.getImageLink() != null) {
            bundledContent.putMap(EVENT_IMAGE, bundleImageSet(content.getImageLink()));
        }

        if (content.getCta() != null) {
            bundledContent.putMap(EVENT_CONTENT_CTA, bundleContentLink(content.getCta()));
        }

        return bundledContent;
    }

    static Content unbundleContent(final ReadableMap bundledContent) {
        final Content content = new Content();
        content.title = getNullableField(bundledContent, EVENT_CONTENT_TITLE);
        content.contentString = getNullableField(bundledContent, EVENT_CONTENT_TEXT);
        if (bundledContent.hasKey(EVENT_IMAGE)) {
            final List<ImageSet> images = new ArrayList<ImageSet>();
            ReadableMap imageSet = bundledContent.getMap(EVENT_IMAGE);
            images.add(unbundleImageSet(imageSet));
            content.setImages_links(images);
        }
        if (bundledContent.hasKey(EVENT_CONTENT_CTA)) {
            ReadableMap bundledCta = bundledContent.getMap(EVENT_CONTENT_CTA);
            if (bundledCta != null) {
                content.setCta(unbundleContentLink(bundledCta));
            }
        }
        return content;
    }

    static WritableArray bundleNotificationHistory(final List<HistoryItem> history) {
        final WritableArray bundledHistory = new WritableNativeArray();

        for (HistoryItem item : history) {
            final WritableMap bundledItem = bundleHistoryItem(item);
            bundledHistory.pushMap(bundledItem);
        }

        return bundledHistory;
    }

    static WritableMap bundleFeedback(final Feedback feedback) {
        final WritableMap bundledFeedback = new WritableNativeMap();

        bundledFeedback.putString(EVENT_FEEDBACK_QUESTION, feedback.question);

        String feedbackB64 = null;
        try {
            feedbackB64 = feedbackToB64(feedback);
        } catch (Exception e) {
            Log.e(TAG, "feedback encoding error", e);
        }
        if (feedbackB64 != null) {
            bundledFeedback.putString(EVENT_FEEDBACK_ID, feedbackB64);
        }

        return bundledFeedback;
    }

    @Nullable
    static Feedback unbundleFeedback(final ReadableMap bundledFeedback) {
        Feedback feedback = null;
        if (bundledFeedback.hasKey(EVENT_FEEDBACK_ID)) {
            try {
                feedback = feedbackFromB64(bundledFeedback.getString(EVENT_FEEDBACK_ID));
            } catch (Exception e) {
                Log.e(TAG, "Could NOT parse Feedback");
            }
        } else {
            Log.e(TAG, "Could NOT parse Feedback");
        }
        return feedback;
    }

    static WritableMap bundleCustomJson(final CustomJSON customJson) {
        return toWritableMap(customJson.content);
    }

    private static WritableMap bundleHistoryItem(final HistoryItem item) {
        final WritableMap itemMap = new WritableNativeMap();

        itemMap.putBoolean(NOTIFICATION_HISTORY_READ, item.read);
        itemMap.putDouble(NOTIFICATION_HISTORY_TIMESTAMP, item.timestamp);
        itemMap.putBoolean(NOTIFICATION_HISTORY_IS_NEW, item.isNew);

        try {
            itemMap.putString(EVENT_TRACKING_INFO, bundleTrackingInfo(item.trackingInfo));
        } catch (Exception e) {
            Log.e(TAG, "historyItem encoding error", e);
        }
        itemMap.putString(EVENT_CONTENT_MESSAGE, item.reaction.notificationMessage);

        if (item.reaction instanceof SimpleNotification) {
            itemMap.putString(EVENT_TYPE, EVENT_TYPE_SIMPLE);
        } else if (item.reaction instanceof Content) {
            Content content = (Content) item.reaction;
            itemMap.putMap(NOTIFICATION_HISTORY_CONTENT, bundleContent(content));
            itemMap.putString(EVENT_TYPE, EVENT_TYPE_CONTENT);
        } else if (item.reaction instanceof Feedback) {
            Feedback feedback = (Feedback) item.reaction;
            itemMap.putMap(NOTIFICATION_HISTORY_CONTENT, bundleFeedback(feedback));
            itemMap.putString(EVENT_TYPE, EVENT_TYPE_FEEDBACK);
        } else if (item.reaction instanceof Coupon) {
            Coupon coupon = (Coupon) item.reaction;
            itemMap.putMap(NOTIFICATION_HISTORY_CONTENT, bundleCoupon(coupon));
            itemMap.putString(EVENT_TYPE, EVENT_TYPE_COUPON);
        } else if (item.reaction instanceof CustomJSON) {
            CustomJSON customJson = (CustomJSON) item.reaction;
            itemMap.putMap(NOTIFICATION_HISTORY_CONTENT, bundleCustomJson(customJson));
            itemMap.putString(EVENT_TYPE, EVENT_TYPE_CUSTOM_JSON);
        }

        return itemMap;
    }

    private static WritableMap bundleImageSet(final ImageSet imageSet) {
        final WritableMap image = new WritableNativeMap();
        image.putString(EVENT_IMAGE_FULL_SIZE, imageSet.getFullSize());
        image.putString(EVENT_IMAGE_SQUARE_SIZE, imageSet.getSmallSize());

        return image;
    }

    private static ImageSet unbundleImageSet(final ReadableMap bundledImage) {
        final ImageSet imageSet = new ImageSet();
        imageSet.setFullSize(getNullableField(bundledImage, EVENT_IMAGE_FULL_SIZE));
        imageSet.setSmallSize(getNullableField(bundledImage, EVENT_IMAGE_SQUARE_SIZE));
        return imageSet;
    }

    private static WritableMap bundleContentLink(final ContentLink cta) {
        final WritableMap contentLink = new WritableNativeMap();
        contentLink.putString(EVENT_CONTENT_CTA_LABEL, cta.label);
        contentLink.putString(EVENT_CONTENT_CTA_URL, cta.url);
        return contentLink;
    }

    private static ContentLink unbundleContentLink(final ReadableMap bundledCta) {
        return new ContentLink(
                getNullableField(bundledCta, EVENT_CONTENT_CTA_LABEL),
                getNullableField(bundledCta, EVENT_CONTENT_CTA_URL));
    }

    private static String getNullableField(final ReadableMap map, final String key) {
        if (map.hasKey(key) && map.getString(key) != null) {
            return map.getString(key);
        }
        return null;
    }

    private static Feedback feedbackFromB64(final String base64) throws Exception {
        Feedback feedback;
        final Parcel parcel = Parcel.obtain();
        try {
            final ByteArrayOutputStream byteBuffer = new ByteArrayOutputStream();
            final byte[] buffer = new byte[1024];
            final GZIPInputStream zis = new GZIPInputStream(new ByteArrayInputStream(Base64.decode(base64, 0)));
            int len;
            while ((len = zis.read(buffer)) != -1) {
                byteBuffer.write(buffer, 0, len);
            }
            zis.close();
            parcel.unmarshall(byteBuffer.toByteArray(), 0, byteBuffer.size());
            parcel.setDataPosition(0);

            feedback = Feedback.CREATOR.createFromParcel(parcel);
        } finally {
            parcel.recycle();
        }

        return feedback;
    }

    private static String feedbackToB64(final Feedback feedback) throws Exception {
        String base64;

        final Parcel parcel = Parcel.obtain();
        try {
            feedback.writeToParcel(parcel, Parcelable.PARCELABLE_WRITE_RETURN_VALUE);
            final ByteArrayOutputStream bos = new ByteArrayOutputStream();
            final GZIPOutputStream zos = new GZIPOutputStream(new BufferedOutputStream(bos));
            zos.write(parcel.marshall());
            zos.close();
            base64 = Base64.encodeToString(bos.toByteArray(), 0);
        } finally {
            parcel.recycle();
        }

        return base64;
    }

    // ReadableArray Utils
    private static JSONArray toJSONArray(final ReadableArray readableArray) throws JSONException {
        JSONArray jsonArray = new JSONArray();

        for (int i = 0; i < readableArray.size(); i++) {
            ReadableType type = readableArray.getType(i);

            switch (type) {
                case Null:
                    jsonArray.put(i, null);
                    break;
                case Boolean:
                    jsonArray.put(i, readableArray.getBoolean(i));
                    break;
                case Number:
                    jsonArray.put(i, readableArray.getDouble(i));
                    break;
                case String:
                    jsonArray.put(i, readableArray.getString(i));
                    break;
                case Map:
                    jsonArray.put(i, toJSONObject(readableArray.getMap(i)));
                    break;
                case Array:
                    jsonArray.put(i, toJSONArray(readableArray.getArray(i)));
                    break;
            }
        }

        return jsonArray;
    }

    private static Object[] toArray(final JSONArray jsonArray) throws JSONException {
        Object[] array = new Object[jsonArray.length()];

        for (int i = 0; i < jsonArray.length(); i++) {
            Object value = jsonArray.get(i);

            if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            if (value instanceof JSONArray) {
                value = toArray((JSONArray) value);
            }

            array[i] = value;
        }

        return array;
    }

    private static Object[] toArray(final ReadableArray readableArray) {
        Object[] array = new Object[readableArray.size()];

        for (int i = 0; i < readableArray.size(); i++) {
            ReadableType type = readableArray.getType(i);

            switch (type) {
                case Null:
                    array[i] = null;
                    break;
                case Boolean:
                    array[i] = readableArray.getBoolean(i);
                    break;
                case Number:
                    array[i] = readableArray.getDouble(i);
                    break;
                case String:
                    array[i] = readableArray.getString(i);
                    break;
                case Map:
                    array[i] = toMap(readableArray.getMap(i));
                    break;
                case Array:
                    array[i] = toArray(readableArray.getArray(i));
                    break;
            }
        }

        return array;
    }

    private static WritableArray toWritableArray(final Object[] array) {
        WritableArray writableArray = Arguments.createArray();

        for (int i = 0; i < array.length; i++) {
            Object value = array[i];

            if (value == null) {
                writableArray.pushNull();
            }
            if (value instanceof Boolean) {
                writableArray.pushBoolean((Boolean) value);
            }
            if (value instanceof Double) {
                writableArray.pushDouble((Double) value);
            }
            if (value instanceof Integer) {
                writableArray.pushInt((Integer) value);
            }
            if (value instanceof String) {
                writableArray.pushString((String) value);
            }
            if (value instanceof Map) {
                writableArray.pushMap(toWritableMap((Map<String, Object>) value));
            }
            if (value != null && value.getClass().isArray()) {
                writableArray.pushArray(toWritableArray((Object[]) value));
            }
        }

        return writableArray;
    }

    // ReactNative Map Utils
    static JSONObject toJSONObject(final ReadableMap readableMap) throws JSONException {
        JSONObject jsonObject = new JSONObject();

        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();

        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();
            ReadableType type = readableMap.getType(key);

            switch (type) {
                case Null:
                    jsonObject.put(key, null);
                    break;
                case Boolean:
                    jsonObject.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    jsonObject.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    jsonObject.put(key, readableMap.getString(key));
                    break;
                case Map:
                    jsonObject.put(key, toJSONObject(readableMap.getMap(key)));
                    break;
                case Array:
                    jsonObject.put(key, toJSONArray(readableMap.getArray(key)));
                    break;
            }
        }

        return jsonObject;
    }

    static Map<String, Object> toMap(final JSONObject jsonObject) throws JSONException {
        Map<String, Object> map = new HashMap<String, Object>();
        Iterator<String> iterator = jsonObject.keys();

        while (iterator.hasNext()) {
            String key = iterator.next();
            Object value = jsonObject.get(key);

            if (value instanceof JSONObject) {
                value = toMap((JSONObject) value);
            }
            if (value instanceof JSONArray) {
                value = toArray((JSONArray) value);
            }

            map.put(key, value);
        }

        return map;
    }

    static Map<String, Object> toMap(final ReadableMap readableMap) {
        Map<String, Object> map = new HashMap<String, Object>();
        ReadableMapKeySetIterator iterator = readableMap.keySetIterator();

        while (iterator.hasNextKey()) {
            String key = iterator.nextKey();
            ReadableType type = readableMap.getType(key);

            switch (type) {
                case Null:
                    map.put(key, null);
                    break;
                case Boolean:
                    map.put(key, readableMap.getBoolean(key));
                    break;
                case Number:
                    map.put(key, readableMap.getDouble(key));
                    break;
                case String:
                    map.put(key, readableMap.getString(key));
                    break;
                case Map:
                    map.put(key, toMap(readableMap.getMap(key)));
                    break;
                case Array:
                    map.put(key, toArray(readableMap.getArray(key)));
                    break;
            }
        }

        return map;
    }

    static WritableMap toWritableMap(final Map<String, Object> map) {
        WritableMap writableMap = Arguments.createMap();

        for (Object o : map.entrySet()) {
            Map.Entry pair = (Map.Entry) o;
            Object value = pair.getValue();

            if (value == null) {
                writableMap.putNull((String) pair.getKey());
            } else if (value instanceof Boolean) {
                writableMap.putBoolean((String) pair.getKey(), (Boolean) value);
            } else if (value instanceof Double) {
                writableMap.putDouble((String) pair.getKey(), (Double) value);
            } else if (value instanceof Integer) {
                writableMap.putInt((String) pair.getKey(), (Integer) value);
            } else if (value instanceof String) {
                writableMap.putString((String) pair.getKey(), (String) value);
            } else if (value instanceof Map) {
                writableMap.putMap((String) pair.getKey(), toWritableMap((Map<String, Object>) value));
            } else if (value.getClass().isArray()) {
                writableMap.putArray((String) pair.getKey(), toWritableArray((Object[]) value));
            }

        }

        return writableMap;
    }
}

File: ios/RNNearItUtils.m
//
//  RNNearItUtils.m
//  RNNearIt
//
//  Created by Federico Boschini <federico@nearit.com> on 10/06/2019.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import "RNNearItUtils.h"

#define TAG @"RNNearItUtils"

@implementation RNNearItUtils

// NITTrackingInfo

+ (NSString* _Nullable)bundleTrackingInfo:(NITTrackingInfo* _Nullable) trackingInfo
{
    NSString* trackingInfoB64;
    if (trackingInfo) {
        NSData* trackingInfoData = [NSKeyedArchiver archivedDataWithRootObject:trackingInfo];
        trackingInfoB64 = [trackingInfoData base64EncodedStringWithOptions:0];
    }
    
    return trackingInfoB64;
}

+ (NITTrackingInfo* _Nullable)unbundleTrackingInfo:(NSString * _Nullable)bundledTrackingInfo
{
    NSData* trackingInfoData = [[NSData alloc] initWithBase64EncodedString:bundledTrackingInfo
                                                                   options:NSDataBase64DecodingIgnoreUnknownCharacters];
    
    NITTrackingInfo *trackingInfo = [NSKeyedUnarchiver unarchiveObjectWithData:trackingInfoData];
    return trackingInfo;
}

// NITCoupon

+ (NSDictionary* _Nullable)bundleNITCoupon:(NITCoupon* _Nonnull) coupon
{
    NSMutableDictionary* couponDictionary = [[NSMutableDictionary alloc] init];
    [couponDictionary setObject:(coupon.title ? coupon.title : [NSNull null])
                         forKey:EVENT_CONTENT_TITLE];
    [couponDictionary setObject:(coupon.couponDescription ? coupon.couponDescription : [NSNull null])
                         forKey:EVENT_COUPON_DESCRIPTION];
    [couponDictionary setObject:(coupon.value ? coupon.value : [NSNull null])
                         forKey:EVENT_COUPON_VALUE];
    [couponDictionary setObject:(coupon.expiresAt ? coupon.expiresAt : [NSNull null])
                         forKey:EVENT_COUPON_EXPIRES_AT];
    [couponDictionary setObject:(coupon.redeemableFrom ? coupon.redeemableFrom : [NSNull null])
                         forKey:EVENT_COUPON_REDEEMABLE_FROM];
    [couponDictionary setObject:(coupon.serial ? coupon.serial : [NSNull null])
                         forKey:EVENT_COUPON_SERIAL];
    [couponDictionary setObject:(coupon.claimedAt ? coupon.claimedAt : [NSNull null])
                         forKey:EVENT_COUPON_CLAIMED_AT];
    [couponDictionary setObject:(coupon.redeemedAt ? coupon.redeemedAt : [NSNull null])
                         forKey:EVENT_COUPON_REDEEMED_AT];
    
    if (coupon.icon) {
        if (coupon.icon.url || coupon.icon.smallSizeURL) {
            [couponDictionary setObject:[self bundleNITImage:coupon.icon] forKey:EVENT_IMAGE];
        }
    }
    
    NSData* couponData = [NSKeyedArchiver archivedDataWithRootObject:coupon];
    NSString* couponB64 = [couponData base64EncodedStringWithOptions:0];
    [couponDictionary setObject:couponB64 forKey:@"couponData"];
    
    return couponDictionary;
}

+ (NITCoupon* _Nullable)unbundleNITCoupon:(NSDictionary* _Nonnull)bundledCoupon
{
    NSString* couponString = [bundledCoupon objectForKey:@"couponData"];
    NSData* couponData = [[NSData alloc] initWithBase64EncodedString:couponString
                                                             options:NSDataBase64DecodingIgnoreUnknownCharacters];
    NITCoupon *coupon = [NSKeyedUnarchiver unarchiveObjectWithData:couponData];
    return coupon;
}

// NITContent

+ (NSDictionary* _Nullable)bundleNITContent:(NITContent * _Nonnull) content
{
    NSString* title = [content title];
    if (!title) {
        title = [NSNull null];
    }
    
    NSString* text = [content content];
    if (!text) {
        text = [NSNull null];
    }
    
    id image;
    if (content.image) {
        image = [self bundleNITImage:content.image];
    } else {
        image = [NSNull null];
    }
    
    id cta;
    if (content.link) {
        cta = [self bundleNITContentLink:content.link];
    } else {
        cta = [NSNull null];
    }
    
    NSDictionary* bundledContent = @{
                                     EVENT_CONTENT_TITLE:title,
                                     EVENT_CONTENT_TEXT:text,
                                     EVENT_IMAGE:image,
                                     EVENT_CONTENT_CTA:cta
                                     };
    
    return bundledContent;
}

+ (NITContent* _Nullable)unbundleNITContent:(NSDictionary * _Nonnull)bundledContent
{
    NITContent* content = [[NITContent alloc] init];
    content.title = [bundledContent objectForKey:EVENT_CONTENT_TITLE];
    content.content = [bundledContent objectForKey:EVENT_CONTENT_TEXT];
    content.images = @[[self unbundleNITImage: [bundledContent objectForKey:EVENT_IMAGE]]];
    content.internalLink = [bundledContent objectForKey:EVENT_CONTENT_CTA];
    return content;
}

// NITFeedback

+ (NSDictionary* _Nullable)bundleNITFeedback:(NITFeedback * _Nonnull) feedback
{
    NSData* feedbackData = [NSKeyedArchiver archivedDataWithRootObject:feedback];
    NSString* feedbackB64 = [feedbackData base64EncodedStringWithOptions:0];
    
    NSDictionary* bundledFeedback = @{
                                      EVENT_FEEDBACK_ID: feedbackB64,
                                      EVENT_FEEDBACK_QUESTION: [feedback question]
                                      };
    
    return bundledFeedback;
}

+ (NITFeedback* _Nullable)unbundleNITFeedback:(NSDictionary * _Nonnull) bundledFeedback
{
    NSString* feedbackId = [bundledFeedback objectForKey:EVENT_FEEDBACK_ID];
    NSData* feedbackData = [[NSData alloc] initWithBase64EncodedString:feedbackId
                                                               options:NSDataBase64DecodingIgnoreUnknownCharacters];
    
    NITFeedback *feedback = [NSKeyedUnarchiver unarchiveObjectWithData:feedbackData];
    feedback.question = [bundledFeedback objectForKey:EVENT_FEEDBACK_QUESTION];
    return feedback;
}

// NITCustomJSON

+ (NSDictionary* _Nullable)bundleNITCustomJSON:(NITCustomJSON* _Nonnull) custom
{
    return @{[custom content]};;
}

// NITImage

- (NSDictionary* _Nullable)bundleNITImage:(NITImage* _Nonnull)image
{
    NSData* imageData = [NSKeyedArchiver archivedDataWithRootObject:image];
    NSString* imageB64 = [imageData base64EncodedStringWithOptions:0];
    return @{
             @"imageData": imageB64,
             EVENT_IMAGE_FULL_SIZE: (image.url ? [image.url absoluteString] : [NSNull null]),
             EVENT_IMAGE_SQUARE_SIZE: (image.smallSizeURL ? [image.smallSizeURL absoluteString] : [NSNull null])
             };
}

- (NITImage* _Nullable)unbundleNITImage:(NSDictionary* _Nonnull)bundledImage
{
    NITImage* image = [[NITImage alloc] init];
    if ([bundledImage objectForKey:@"imageData"]) {
        NSData* imageData = [[NSData alloc] initWithBase64EncodedString:[bundledImage objectForKey:@"imageData"]
                                                                options:NSDataBase64DecodingIgnoreUnknownCharacters];
        image = [NSKeyedUnarchiver unarchiveObjectWithData:imageData];
    }
    return image;
}

// NITContentLink

- (NSDictionary* _Nullable)bundleNITContentLink:(NITContentLink* _Nonnull)cta
{
    return @{
             EVENT_CONTENT_CTA_LABEL: cta.label,
             EVENT_CONTENT_CTA_URL: [cta.url absoluteString]
             };
}

- (NITContentLink* _Nullable)unbundleNITContentLink:(NSDictionary* _Nonnull)bundledCta
{
    
}

// NITHistoryItem

- (NSDictionary* _Nullable)bundleNITHistoryItem:(NITHistoryItem* _Nonnull) item
{
    NSMutableDictionary* historyDictionary = [[NSMutableDictionary alloc] init];
    
    NSNumber *read = [NSNumber numberWithBool:item.read];
    NSNumber *timestamp = [NSNumber numberWithDouble:item.timestamp];
    NSString *bundledTrackingInfo = [self bundleTrackingInfo:item.trackingInfo];
    NSNumber *isNew = [NSNumber numberWithBool:item.isNew];
    
    [historyDictionary setObject:read forKey:NOTIFICATION_HISTORY_READ];
    [historyDictionary setObject:timestamp forKey:NOTIFICATION_HISTORY_TIMESTAMP];
    [historyDictionary setObject:isNew forKey:NOTIFICATION_HISTORY_IS_NEW];
    [historyDictionary setObject:(bundledTrackingInfo ? bundledTrackingInfo : [NSNull null]) forKey:EVENT_TRACKING_INFO];
    
    NSString* message = item.reactionBundle.notificationMessage;
    if (!message) {
        message = [NSNull null];
    }
    [historyDictionary setObject:message forKey:EVENT_CONTENT_MESSAGE];
    
    if ([item.reactionBundle isKindOfClass:[NITSimpleNotification class]]) {
        
        [historyDictionary setObject:EVENT_TYPE_SIMPLE forKey:EVENT_TYPE];
        
    } else if ([item.reactionBundle isKindOfClass:[NITContent class]]) {
        
        [historyDictionary setObject:EVENT_TYPE_CONTENT forKey:EVENT_TYPE];
        
        NITContent *nearContent = (NITContent*)item.reactionBundle;
        NSDictionary* content = [self bundleNITContent:nearContent];
        [historyDictionary setObject:content forKey:NOTIFICATION_HISTORY_CONTENT];
        
    } else if ([item.reactionBundle isKindOfClass:[NITFeedback class]]) {
        
        [historyDictionary setObject:EVENT_TYPE_FEEDBACK forKey:EVENT_TYPE];
        
        NITFeedback* nearFeedback = (NITFeedback*)item.reactionBundle;
        NSDictionary* feedback = [self bundleNITFeedback:nearFeedback];
        [historyDictionary setObject:feedback forKey:NOTIFICATION_HISTORY_CONTENT];
        
    } else if ([item.reactionBundle isKindOfClass:[NITCoupon class]]) {
        
        [historyDictionary setObject:EVENT_TYPE_COUPON forKey:EVENT_TYPE];
        
        
    } else if ([item.reactionBundle isKindOfClass:[NITCustomJSON class]]) {
        
        [historyDictionary setObject:EVENT_TYPE_CUSTOM_JSON forKey:EVENT_TYPE];
        
        NITCustomJSON *nearCustom = (NITCustomJSON*)item.reactionBundle;
        NSDictionary* custom = [self bundleNITCustomJSON:nearCustom];
        [historyDictionary setObject:custom forKey:NOTIFICATION_HISTORY_CONTENT];
    }
    
    return historyDictionary;
}

@end
